examples/magic_square.py

```diff
@@ -0,0 +1,474 @@
+    elif mermin_row == 2:
+        alice_circuit = cirq.Circuit(
+            cirq.CZ(*q),
+            cirq.Moment(cirq.H.on_each(*q, *m)),
+            cirq.CZ.on_each(*zip(m, q)),
+            cirq.H.on_each(*m),
+            cirq.Moment(cirq.M(*m, key="alice")),
+        )
+
+    q = bob_qubits[1:3]  # data qubits
+    m = (bob_qubits[0], bob_qubits[3])  # measure qubits
+    if mermin_col == 0:
+        bob_circuit = cirq.Circuit(cirq.H(q[0]), cirq.M(*q, key="bob"))
+    elif mermin_col == 1:
+        bob_circuit = cirq.Circuit(cirq.H(q[1]), cirq.M(*q, key="bob"))
+    elif mermin_col == 2:
+        bob_circuit = cirq.Circuit(
+            cirq.H(q[0]),
+            cirq.CZ(*q),
+            cirq.Moment(cirq.H.on_each(*q, *m)),
+            cirq.CZ.on_each(*zip(m, q)),
+            cirq.H.on_each(*m),
+            cirq.Moment(cirq.M(*m, key="bob")),
+        )
+
+    last_moment_circuit = cirq.Circuit(cirq.Moment(alice_circuit[-1] + bob_circuit[-1]))
+    if mermin_row == 0:
+        circuit = bob_circuit[:-1] + last_moment_circuit
+    elif mermin_row == 1:
+        if mermin_col == 2:
+            circuit = bob_circuit[:-2] + alice_circuit.zip(bob_circuit[-2:])
+        else:
+            circuit = alice_circuit[:-1].zip(bob_circuit[:-1]) + last_moment_circuit
+    elif mermin_row == 2:
+        if mermin_col == 2:
+            circuit = (
+                cirq.Circuit(bob_circuit[0])
+                + alice_circuit[:-1].zip(bob_circuit[1:-1])
+                + last_moment_circuit
+            )
+        else:
+            circuit = (
+                alice_circuit[0:3]
+                + cirq.Circuit(cirq.Moment(alice_circuit[3] + bob_circuit[0]))
+                + last_moment_circuit
+            )
+    return circuit
+
+
+def construct_magic_square_circuit(
+    alice_qubits: list[cirq.GridQubit],
+    bob_qubits: list[cirq.GridQubit],
+    mermin_row: int,
+    mermin_col: int,
+    add_dd: bool,
+    dd_scheme: tuple[cirq.Gate, ...] = (cirq.X, cirq.Y, cirq.X, cirq.Y),
+) -> cirq.Circuit:
+    """Construct a circuit to implement the magic square game.
+
+    We use the conventions described in https://en.wikipedia.org/wiki/Quantum_pseudo-telepathy.
+    In particular, the Mermin-Peres table is
+
+     I ⊗ Z | Z ⊗ I | Z ⊗ Z
+     X ⊗ I | I ⊗ X | X ⊗ X
+    -X ⊗ Z |-Z ⊗ X | Y ⊗ Y
+
+    Args:
+        alice_qubits: The line of qubits to use for Alice (measure, data, data, measure).
+        bob_qubits: The line of qubits to use for Bob (measure, data, data, measure).
+        mermin_row: The row of the Mermin-Peres square to measure.
+        mermin_col: The column of the Mermin-Peres square to measure.
+        add_dd: Whether to add dynamical decoupling.
+        dd_scheme: The dynamical decoupling sequence to use if doing DD.
+
+    Returns:
+        A circuit implementing the game.
+    """
+    alice_data_qubits = (alice_qubits[1], alice_qubits[2])
```

Comment:
I'd add

```python
assert len(alice_qubits) == 4, f"Expected 4 qubits, but got {len(alice_qubits)}"
assert len(bob_qubits) == 4, f"Expected 4 qubits, but got {len(bob_qubits)}"
```

to indicate error earlier rather than later. Errors raised earlier are typically more helpful in debugging.

If you need it to work for more qubits, then `in (4, 5)` or `>= 4` are still better than no check at all.